File: Java/1443. Minimum Time to Collect All Apples in a Tree.java
class Solution {
    public int minTime(int n, int[][] edges, List<Boolean> hasApple) {
        Map<Integer, List<Integer>> adjs = new HashMap();
        for (int[] edge: edges) {
            adjs.computeIfAbsent(edge[0], value -> new ArrayList<Integer>()).add(edge[1]);
            adjs.computeIfAbsent(edge[1], value -> new ArrayList<Integer>()).add(edge[0]);
        }

        return dfs(0, -1, hasApple, adjs);
    }

    private int dfs(int node, int parent, List<Boolean> hasApple, Map<Integer, List<Integer>> adjs) {
        if (!adjs.containsKey(node)) {
            return 0;
        }

        int total = 0, child = 0;
        for (int nei: adjs.get(node)) {
            if (nei == parent) {
                continue;
            }
            child = dfs(nei, node, hasApple, adjs);
            if (child > 0 || hasApple.get(nei)) {
                total += child + 2;
            }
        }

        return total;
    }
}

File: Python3/1443. Minimum Time to Collect All Apples in a Tree.py
class Solution:
    def minTime(self, n: int, edges: List[List[int]], hasApple: List[bool]) -> int:
        ans = 0
        adjs = collections.defaultdict(list)

        for p, c in edges:
            adjs[p].append(c)
            adjs[c].append(p)

        def dfs(node = 0, parent = -1):
            if node not in adjs:
                return 0

            total = child = 0
            for nei in adjs[node]:
                if nei == parent:
                    continue

                child = dfs(nei, node)
                if child > 0 or hasApple[nei]:
                    total += child + 2

            return total

        return dfs()
        